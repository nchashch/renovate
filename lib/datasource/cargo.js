const got = require('got');

module.exports = {
  getPkgReleases,
};

async function getPkgReleases(purl) {
  const { fullname: name } = purl;
  logger.trace(`cargo.getPkgReleases(${name})`);
  const baseUrl = 'https://crates.io/api/v1/crates/';
  const crateUrl = baseUrl + name;
  try {
    const res = (await got(crateUrl, {
      json: true,
      retry: 5,
    })).body;
    if (res.crate.name !== name) {
      return null;
    }
    const releases = [];
    for (let i = 0; i < res.versions.length; i += 1) {
      const release = {
        version: res.versions[i].num,
      };
      releases.push(release);
    }
    const result = {
      releases,
    };
    result.repository = res.crate.repository;
    const homepage = res.crate.homepage;
    if (result.repository !== homepage) {
      result.homepage = homepage;
    }
    return result;
  } catch (err) {
    if (err.statusCode === 404 || err.code === 'ENOTFOUND') {
      logger.info({ dependency: name }, `Dependency lookup failure: not found`);
      logger.debug({
        err,
      });
      return null;
    }
    logger.info({ err, name }, 'crates.io lookup failure: Unknown error');
    return null;
  }
}
