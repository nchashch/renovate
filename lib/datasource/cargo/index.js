const got = require('../../util/got');

module.exports = {
  getPkgReleases,
};

async function getPkgReleases({ lookupName }) {
  if (!lookupName) {
    return null;
  }
  const len = lookupName.length;
  let path;
  if (len === 1) {
    path = '1/' + lookupName;
  } else if (len === 2) {
    path = '2/' + lookupName;
  } else if (len === 3) {
    path = '3/' + lookupName[0] + '/' + lookupName;
  } else {
    path =
      lookupName.slice(0, 2) + '/' + lookupName.slice(2, 4) + '/' + lookupName;
  }
  const baseUrl =
    'https://raw.githubusercontent.com/rust-lang/crates.io-index/master/';
  const crateUrl = baseUrl + path;
  try {
    const res = (await got(crateUrl, {
      json: true,
      platform: 'cargo',
    })).body;
    if (!(res && res.crate && res.crate.name && res.versions)) {
      logger.warn({ dependency: lookupName }, `Received invalid crate data`);
      return null;
    }
    const result = {
      releases: [],
    };
    result.releases = res.versions.map(version => {
      const release = {
        version: version.num,
      };
      if (version.yanked) {
        release.isDeprecated = true;
      }
      return release;
    });
    result.sourceUrl = res.crate.repository;
    result.homepage = res.crate.homepage;
    return result;
  } catch (err) {
    if (err.statusCode === 404 || err.code === 'ENOTFOUND') {
      logger.info({ lookupName }, `Dependency lookup failure: not found`);
      logger.debug(
        {
          err,
        },
        'Crate lookup error'
      );
      return null;
    }
    if (
      err.statusCode === 429 ||
      (err.statusCode > 500 && err.statusCode < 600)
    ) {
      logger.warn({ lookupName, err }, `cargo crates.io registry failure`);
      throw new Error('registry-failure');
    }
    logger.warn(
      { err, lookupName },
      'cargo crates.io lookup failure: Unknown error'
    );
    return null;
  }
}
