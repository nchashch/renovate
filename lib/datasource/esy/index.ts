import { logger } from '../../logger';
import got from '../../util/got';
import { PkgReleaseConfig, ReleaseResult, Release } from '../common';

export async function getPkgReleases({
  lookupName,
}: PkgReleaseConfig): Promise<ReleaseResult> {
  const path: string = `/repos/ocaml/opam-repository/contents/packages/${lookupName}/`;
  const baseUrl: string = 'https://api.github.com';
  const packageUrl: string = baseUrl + path;
  console.log(packageUrl);
  let res: any = await got(packageUrl, {
    hostType: 'esy',
    json: true,
  });
  // TODO: Read homepage, name, and maybe other metadata from /packages/${lookupName}/${lookupName.1.2.3}/opam
  // where 1.2.3 can be any existing version, check if there is a mechanism for deprication
  const versions: string[] = res.body.map(file =>
    file.name.substring(file.name.indexOf('.') + 1)
  );
  const releases: Release[] = versions.map(version => ({ version }));
  const result: ReleaseResult = { releases };
  return result;
}
