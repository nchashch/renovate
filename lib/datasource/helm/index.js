const yaml = require('js-yaml');
const got = require('../../util/got');

module.exports = {
  getPkgReleases,
};

async function getPkgReleases({ lookupName, repository }) {
  if (!lookupName) {
    logger.warn(`lookupName was not provided to getPkgReleases`);
    return null;
  }
  if (!repository) {
    logger.warn(`repository was not provided to getPkgReleases`);
    return null;
  }
  try {
    const res = await got('index.yaml', { baseUrl: repository });
    if (!res || !res.body) {
      logger.warn(
        { dependency: lookupName },
        `Received invalid index.yaml from ${repository}`
      );
      return null;
    }
    try {
      const doc = yaml.safeLoad(res.body);
      if (!doc) {
        logger.warn(
          { dependency: lookupName },
          `Failed to parse index.yaml from ${repository}`
        );
        return null;
      }
      let releases = doc.entries[lookupName];
      if (!releases) {
        logger.warn(
          { dependency: lookupName },
          `Entry ${lookupName} doesn't exist in index.yaml from ${repository}`
        );
        return null;
      }
      releases = releases.map(entry => ({
        version: entry.version,
      }));
      return {
        releases,
      };
    } catch (yamlErr) {
      return null;
    }
  } catch (gotErr) {
    if (gotErr.statusCode === 404 || gotErr.code === 'ENOTFOUND') {
      logger.info({ lookupName }, `Dependency lookup failure: not found`);
      logger.debug({ gotErr }, 'index.yaml lookup error');
      return null;
    }
    if (
      gotErr.statusCode === 429 ||
      (gotErr.statusCode > 500 && gotErr.statusCode < 600)
    ) {
      logger.warn({ lookupName, gotErr }, `${repository} server error`);
      throw new Error('registry-failure');
    }
    logger.warn(
      { gotErr, lookupName },
      `${repository} lookup failure: Unknown error`
    );
    return null;
  }
}
