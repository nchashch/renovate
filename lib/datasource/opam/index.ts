import { logger } from '../../logger';
import got from '../../util/got';
import { PkgReleaseConfig, ReleaseResult, Release } from '../common';
import { parseDepName } from '../../manager/esy/extract';

export async function getPkgReleases({
  lookupName,
}: PkgReleaseConfig): Promise<ReleaseResult> {
  const { depName } = parseDepName(lookupName);
  const cacheNamespace = 'datasource-opam';
  const cacheKey = lookupName;
  const cachedResult = await renovateCache.get<ReleaseResult>(
    cacheNamespace,
    cacheKey
  );
  // istanbul ignore if
  if (cachedResult) {
    return cachedResult;
  }
  const path = `/repos/ocaml/opam-repository/contents/packages/${depName}/`;
  const baseUrl = 'https://api.github.com';
  const packageUrl = baseUrl + path;
  let res: any;
  try {
    res = await got(packageUrl, {
      hostType: 'esy',
      json: true,
    });
  } catch (err) {
    if (
      err.statusCode === 429 ||
      (err.statusCode >= 500 && err.statusCode < 600)
    ) {
      logger.warn(
        { lookupName, err },
        `opam repository github.com registry failure`
      );
      throw new Error('registry-failure');
    }
    if (err.statusCode === 404 || err.code === 'ENOTFOUND') {
      logger.warn({ lookupName }, `Dependency lookup failure: not found`);
      logger.debug({ err }, 'opam lookup error');
      return null;
    }
    return null;
  }
  const versions: string[] = res.body.map(file =>
    file.name.substring(file.name.indexOf('.') + 1)
  );
  let homepage: string;
  try {
    const packageVersionUrl = baseUrl + path + `${depName}.${versions[0]}/opam`;
    res = await got(packageVersionUrl, {
      hostType: 'esy',
      json: true,
    });
    const opamContent = Buffer.from(res.body.content, 'base64').toString(
      'utf8'
    );
    homepage = extractHomepage(opamContent);
  } catch (err) {}
  const releases: Release[] = versions.map(version => ({ version }));
  const result: ReleaseResult = { name: depName, releases };
  if (homepage) {
    result.homepage = homepage;
  }
  const cacheMinutes = 10;
  await renovateCache.set(cacheNamespace, cacheKey, result, cacheMinutes);
  return result;
}

function extractHomepage(opamContentWithComments) {
  let opamContent = opamContentWithComments;
  opamContent = removeLineComments(opamContent);
  opamContent = removeMultiLineComments(opamContent);
  const homepageIndex = opamContent.indexOf('homepage');
  if (homepageIndex < 0) {
    return null;
  }
  let homepage = opamContent.substring(homepageIndex);
  homepage = homepage.substring('homepage'.length);
  const colonIndex = skip(0, homepage, isSpace);
  if (homepage[colonIndex] !== ':') {
    return null;
  }
  homepage = homepage.substring(colonIndex + 1);
  // According to opam docs https://opam.ocaml.org/doc/1.2/Manual.html
  // strings are enclosed in "double quotes"
  const doubleQuoteIndex = skip(0, homepage, isSpace);
  if (homepage[doubleQuoteIndex] !== '"') {
    return null;
  }
  homepage = homepage.substring(doubleQuoteIndex + 1);
  const secondDoubleQuoteIndex = homepage.indexOf('"');
  if (secondDoubleQuoteIndex < 0) {
    return null;
  }
  homepage = homepage.substring(0, secondDoubleQuoteIndex);
  return homepage;
}

function skip(
  idx: number,
  content: string,
  cond: (s: string) => boolean
): number {
  let i = idx;
  while (i < content.length) {
    if (!cond(content[i])) {
      return i;
    }
    i += 1;
  }
  // istanbul ignore next
  return i;
}

function isSpace(c: string): boolean {
  return /\s/.test(c);
}

// Remove multi-line comments enclosed between (* and *)
function removeMultiLineComments(content: string): string {
  const beginRegExp = /\(\*/;
  const endRegExp = /\*\)/;
  let newContent = content;
  let i = newContent.search(beginRegExp);
  let j = newContent.search(endRegExp);
  while (i !== -1 && j !== -1) {
    j += '*)'.length;
    newContent = newContent.substring(0, i) + newContent.substring(j);
    i = newContent.search(beginRegExp);
    j = newContent.search(endRegExp);
  }
  return newContent;
}

// Remove line comments starting with #
function removeLineComments(content: string): string {
  let newContent = '';
  let comment = false;
  for (let i = 0; i < content.length; i += 1) {
    const c = content[i];
    if (c === '#') {
      comment = true;
    }
    if (comment) {
      if (c === '\n') {
        comment = false;
      }
    }
    if (!comment) {
      newContent += c;
    }
  }
  return newContent;
}
