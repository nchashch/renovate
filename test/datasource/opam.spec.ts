import * as fs from 'fs';

import _got from '../../lib/util/got';
import { getPkgReleases } from '../../lib/datasource/opam';

const got: any = _got;

const versionsResBody = fs.readFileSync(
  'test/datasource/opam/_fixtures/versions-res.json',
  'utf8'
);

const packageResBody = fs.readFileSync(
  'test/datasource/opam/_fixtures/package-res.json',
  'utf8'
);

jest.mock('../../lib/util/got');

describe('datasource/opam', () => {
  describe('getPkgReleases', () => {
    it('returns versions if both requests to GitHub API are successful', async () => {
      const versionsRes = {
        body: JSON.parse(versionsResBody),
      };
      const packageRes = {
        body: JSON.parse(packageResBody),
      };
      got.mockReturnValueOnce(versionsRes).mockReturnValueOnce(packageRes);
      let res = await getPkgReleases({ lookupName: '0install' });
      expect(res).not.toBeNull();
      expect(res).toMatchSnapshot();
    });
  });
});
