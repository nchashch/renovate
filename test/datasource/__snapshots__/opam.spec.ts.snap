// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`datasource/opam getPkgReleases returns versions if both requests to GitHub API are successful 1`] = `
Object {
  "homepage": "http://0install.net/",
  "name": "0install",
  "releases": Array [
    Object {
      "version": "2.10",
    },
    Object {
      "version": "2.11",
    },
    Object {
      "version": "2.12.1",
    },
    Object {
      "version": "2.12.3",
    },
    Object {
      "version": "2.12",
    },
    Object {
      "version": "2.14.1",
    },
    Object {
      "version": "2.14",
    },
    Object {
      "version": "2.6.2",
    },
    Object {
      "version": "2.8",
    },
    Object {
      "version": "2.9.1",
    },
  ],
}
`;
